pupil_src/shared_modules/annotations.py

```diff
@@ -10,220 +10,297 @@
-                return lambda _: self.fire_annotation(e_name)
-
-            def make_remove(e_name, hotkey):
-                return lambda: self.remove_annotation((e_name, hotkey))
-
-            thumb = ui.Thumb(
-                e_name,
-                setter=make_fire(e_name, hotkey),
-                getter=lambda: False,
-                label=hotkey,
-                hotkey=hotkey,
-            )
-            self.buttons.append(thumb)
-            self.g_pool.quickbar.append(thumb)
-            self.sub_menu.append(
-                ui.Button(e_name + " <" + hotkey + ">", make_remove(e_name, hotkey))
-            )
+        self.customize_menu()
+        self.menu.append(
+            ui.Text_Input("_new_annotation_label", self, label="New Label")
+        )
+        self.menu.append(
+            ui.Text_Input("_new_annotation_hotkey", self, label="New Hotkey")
+        )
+        self.menu.append(ui.Button("Add Annotation Type", self._on_add_annotation_clicked))
+        self._annotation_list_menu = ui.Growing_Menu(
+            "Annotation List - Click to Remove"
+        )
+        self.menu.append(self._annotation_list_menu)
+        self._create_initial_annotation_list()
+
+    def _create_initial_annotation_list(self):
+        for label, hotkey in self._initial_annotation_definitions:
+            self._add_annotation_definition(label, hotkey)
+
+    @abstractmethod
+    def customize_menu(self):
+        pass
 
     def deinit_ui(self):
-        self.menu.remove(self.sub_menu)
-        self.sub_menu = None
+        self._clear_button_quickbars()
         self.remove_menu()
-        if self.buttons:
-            for b in self.buttons:
-                self.g_pool.quickbar.remove(b)
-            self.buttons = []
-
-    def append_annotation(self):
-        self.annotation_definitions.append(
-            (self.new_annotation_name, self.new_annotation_hotkey)
+
+    def _clear_button_quickbars(self):
+        # only call this from deinit_ui()
+        for buttons in self._definition_to_buttons.values():
+            self._remove_button_quickbar(buttons.quickbar)
+
+    @abstractmethod
+    def fire_annotation(self, annotation_definition):
+        pass
+
+    def _on_add_annotation_clicked(self):
+        # new_annotation_label and hotkey are set by the ui input fields
+        self._add_annotation_definition(
+            self._new_annotation_label, self._new_annotation_hotkey
         )
-        self.update_buttons()
 
-    def remove_annotation(self, annotation):
-        try:
-            self.annotation_definitions.remove(annotation)
-        except ValueError:
-            print(annotation, self.annotation_definitions)
-        self.update_buttons()
+    def _add_annotation_definition(self, annotation_label, hotkey):
+        annotation_definition = AnnotationDefinition(
+            label=annotation_label, hotkey=hotkey
+        )
+        if annotation_definition in self._definition_to_buttons:
+            return
```

Comment:
Can this be triggered by the user? If yes, there should be a warning.

Reply:
Right, thanks!